### ISSUE TITLE
Improve lookupRecordType in names

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4283]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)| @fijal reported|
|-|-|
|Trac ID|trac#4283|
|Type|enhancement|
|Created|2010-02-19 21:05:09Z|
|Branch|https://github.com/twisted/twisted/tree/unglobal-lookuprecordtype-4283|

lookupRecordType in names is using globals() right now and a lookup of a constructed string. Would be much better (for pypy especially) to have a dictionary name -> record type instead.

<details><summary>Searchable metadata</summary>

```
trac-id__4283 4283
type__enhancement enhancement
reporter__fijal fijal
priority__normal normal
milestone__ 
branch__branches_unglobal_lookuprecordtype_4283 branches/unglobal-lookuprecordtype-4283
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__pypy pypy
time__1266613509000000 1266613509000000
changetime__1267454156000000 1267454156000000
version__None None
owner__ 
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [28549]) Branching to 'unglobal-lookuprecordtype-4283'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [28550]) Use a specially constructed lookup dict instead of globals()

refs [#4283](https://github.com/twisted/twisted/issues/4283)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

Suggestion implemented in the branch.  I _think_ the existing tests cover this code pretty well; the ones handling actually dns requests all need to call it, anyway.  The only tests I could think to add would be ones that call `lookupRecordType` directly and assert that the returned value is exactly the expected class object.  There'd need to be about 30 of these, and they amount to just a duplication of the information present in the `Record_*` class names.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess set owner to @exarkun|
|-|-|

Thanks for the branch, exarkun. The implementation looks strictly less scary than what's in trunk. A few comments:
 * There's some trailing whitespace in the file 
 * In lookupRecordType: @param and @return need @types
 * class Message needs a docstring.
 * There aren't any tests along the code path where ```lookupRecordType``` returns ```None``` (although I admit it doesn't look like a very interesting case because ```parseRecords``` gets a ```num``` of 0 several times in the tests already).
 * I had to poke around and prove to myself that the ```del name``` line made sense. Maybe warrants a comment?
 * [builds](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/unglobal-lookuprecordtype-4283) are happy.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [28578]) Address review comments

  1. Clean up trailing whitespace in dns.py
  1. Add more epytext markup to `lookupRecordType`
  1. Add a class docstring for `Message`
  1. Add a test for `lookupRecordType` returning `None`
  1. Add a comment about the `del name` in the `Message` class scope

refs [#4283](https://github.com/twisted/twisted/issues/4283)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [28579]) Use the right name

refs [#4283](https://github.com/twisted/twisted/issues/4283)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

Thanks for the awesome review!  All points addressed, I think.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess set owner to @exarkun|
|-|-|

 * test_dns.py needs a copyright bump
 * class MessageTestCase needs a docstring
Other than that, looks good to merge. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [28588]) Address review feedback

refs [#4283](https://github.com/twisted/twisted/issues/4283)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [28590]) Merge unglobal-lookuprecordtype-4283

Author: exarkun
Reviewer: jesstess
Fixes: [#4283](https://github.com/twisted/twisted/issues/4283)

Stop using `globals()` in Twisted Names to look up record classes by
type.  Instead, construct a dictionary just holding this information
which can be keyed without constructing specially formatted strings.
This has nice performance consequences on PyPy.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
